src/socket_listener_berkeley.c

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <sys/types.h>
+#include <sys/unistd.h>
+#include <sys/socket.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include "azure_c_shared_utility/gballoc.h"
+#include "azure_uamqp_c/socket_listener.h"
+#include "azure_c_shared_utility/optimize_size.h"
+#include "azure_c_shared_utility/socketio.h"
+
+typedef struct SOCKET_LISTENER_INSTANCE_TAG
+{
+	int port;
+	int socket;
+	ON_SOCKET_ACCEPTED on_socket_accepted;
+	void* callback_context;
+} SOCKET_LISTENER_INSTANCE;
+
+SOCKET_LISTENER_HANDLE socketlistener_create(int port)
+{
+	SOCKET_LISTENER_INSTANCE* result = (SOCKET_LISTENER_INSTANCE*)malloc(sizeof(SOCKET_LISTENER_INSTANCE));
+	if (result != NULL)
+	{
+		result->port = port;
+		result->on_socket_accepted = NULL;
+		result->callback_context = NULL;
+	}
+
+	return (SOCKET_LISTENER_HANDLE)result;
+}
+
+void socketlistener_destroy(SOCKET_LISTENER_HANDLE socket_listener)
+{
+	if (socket_listener != NULL)
+	{
+		socketlistener_stop(socket_listener);
+		free(socket_listener);
+	}
+}
+
+int socketlistener_start(SOCKET_LISTENER_HANDLE socket_listener, ON_SOCKET_ACCEPTED on_socket_accepted, void* callback_context)
+{
+	int result;
+
+	if (socket_listener == NULL)
+	{
+		result = __FAILURE__;
+	}
+	else
+	{
+		SOCKET_LISTENER_INSTANCE* socket_listener_instance = (SOCKET_LISTENER_INSTANCE*)socket_listener;
+
+		socket_listener_instance->socket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+		if (socket_listener_instance->socket == -1)
+		{
+			result = __FAILURE__;
+		}
+		else
+		{
+            struct sockaddr_in sa;
+
+			socket_listener_instance->on_socket_accepted = on_socket_accepted;
+			socket_listener_instance->callback_context = callback_context;
+
+            sa.sin_family = AF_INET;
+            sa.sin_port = htons(socket_listener_instance->port);
+            sa.sin_addr.s_addr = htonl(INADDR_ANY);
```

Comment:
>sa.sin_addr.s_addr = htonl(INADDR_ANY); [](start = 12, length = 39)

Nitpick, it would be nice to be able to only listen on localhost or a single interface.